### PR TITLE
Add "Enable Prebuilds" to Project Settings – EXP-573

### PR DESCRIPTION
**Pull request by AlexTugarev**
## Description
This should add an explicit "Enable Prebuilds" checkbox to project settings.

| Not enabled | Enabled |
|--------|--------|
|  Screenshot 2023-09-13 at 15 25 54 | Screenshot 2023-09-13 at 15 25 59 |

Pre-existing projects are treated as if prebuilds were enabled, i.e. if new settings aren't present prebuilds are enabled and you'd need to disable them manually. Prebuilds are disabled when new projects are created.

#### Visualizing the prebuild enablement state in the list of all projects
Screenshot 2023-09-13 at 15 22 52

#### New CTA on Project Created page guiding users to the project settings
Screenshot 2023-09-13 at 15 33 27

#### Known issues
- [x] webhooks aren't installed and no errors is shown 

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at 4f80bee</samp>

No summary available (Limit exceeded: required to process 129049 tokens, but only 50000 are allowed per call)

</details>

## Related Issue(s)

Fixes EXP-573

## How to test
Unfortunately, this requires a full manual test of prebuild setups with all supported SCMs.

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - at-enablep552d678423</li>
	<li><b>🔗 URL</b> - <a href="https://at-enablep552d678423.preview.gitpod-dev.com/workspaces" target="_blank">at-enablep552d678423.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - at-enablePrebuilds-settings-gha.16748</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-at-enablep552d678423%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-core-dev" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/server/src/container-module.ts
Comment (geropl) [inline components/server/src/container-module.ts:343]: :heart:

File: components/server/src/projects/projects-service.ts
Comment (geropl) [inline components/server/src/projects/projects-service.ts:239]: How do we educate people that they need to enable Prebuilds now? :thinking:

File: components/gitpod-protocol/src/teams-projects-protocol.ts
Comment (geropl) [inline components/gitpod-protocol/src/teams-projects-protocol.ts:17]: I feel I don't have :100: of the context, but sharing my thought anyway: I somehow expected this to be `prebuild: PrebuildConfig`:
```
type PrebuildConfig {
  enabled?: boolean;
  ...
}
```

**Comment by geropl** [comment]
Can't test yet, it seems, to waiting with the :heavy_check_mark: :upside_down_face:

File: components/gitpod-protocol/src/teams-projects-protocol.ts
Comment (AlexTugarev) [reply components/gitpod-protocol/src/teams-projects-protocol.ts:17]: @geropl, thanks for the comment. I've created this as WIP, but GitHub extension failed to switch to draft mode. 

Can you elaborate on the need for that? Currently in gitpod-protocol, which is also the base for JSON to be stored in DB, it's a flat list of settings. PAPI introduced another structure for settings, which is not perfectly aligned with server/dashboard, e.g. empty state behaves differently. It seems here already have inconsistency to some degree. Not sure, if that's generally known.  
```
type ProjectSettings struct {
	state         protoimpl.MessageState
	sizeCache     protoimpl.SizeCache
	unknownFields protoimpl.UnknownFields

	Prebuild  *PrebuildSettings  `protobuf:"bytes,1,opt,name=prebuild,proto3" json:"prebuild,omitempty"`
	Workspace *WorkspaceSettings `protobuf:"bytes,2,opt,name=workspace,proto3" json:"workspace,omitempty"`
}
```

What's the benefit of the suggested change to protocol?

And where is the justification for the data migration efforts? Sidebar: I'm trying to reduce the need of data migration to zero if possible in this PR.

Comment (geropl) [reply components/gitpod-protocol/src/teams-projects-protocol.ts:17]: I was not aware we have pre-existing data to migrate from.
I just felt encapsulating those options feels cleaner.

File: components/server/src/projects/projects-service.ts
Comment (AlexTugarev) [reply components/server/src/projects/projects-service.ts:239]: 1. After project creation, we render a link to project's settings. 
2. Yes. 

But in general, speaking of educating users: do you think it was commonly known that prebuilds are enabled automatically? 
This is indeed a point we're going to address with updates to design and messages.

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (loujaybee) [inline components/dashboard/src/projects/ProjectSettings.tsx:148]: Thanks, @AlexTugarev !

**thought(blocking):** I assume that clicking enable will install the webhook? If so, that's not super obvious to me as a user. Mentioning the "webhook" at all here feels confusing, could we leave it up to the error? e.g. the user clicks to enable, if incorrect permissions show an error with an action? There feels like too much information that we'd need to contain here, e.g. we'd also need to mention that this overrides the `init` and `before` steps of your `.gitpod.yml`. @gtsiolis -> Any chance we could do a quick mockup and iterate on the copy text for @AlexTugarev? Happy to do a quick sync tomorrow morning if needed to get this unblocked.

Comment (AlexTugarev) [reply components/dashboard/src/projects/ProjectSettings.tsx:148]: > I assume that clicking enable will install the webhook?

yes

> If so, that's not super obvious to me as a user.

that's not a change to current state at all. in fact, you would actually get feedback if it fails. today, if it fails, you wouldn't even notice as everything happens behind the scene and there is no feedback. 

> Mentioning the "webhook" at all here feels confusing, could we leave it up to the error? e.g. the user clicks to enable, if incorrect permissions show an error with an action?

yes. I like that! let's simplify the hint, not mentioning the webhook and improve the error message. 
CTA? which one? I think the most we can do is point out to check write permissions in Gitpod and admin permissions on the project at the SCM provider.

👍🏻 for the sync, but I don't see a blocker her TBH. this is where we need input, thus please fill in the gaps 🙏🏻

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (selfcontained) [inline components/dashboard/src/projects/ProjectSettings.tsx:144]: Can drop the empty `className` here

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (selfcontained) [inline components/dashboard/src/projects/ProjectSettings.tsx:153]: Would be great to use `<InputField>` to handle wrapping the `<SelectWorkspaceClassComponent>` (see [here](https://github.com/gitpod-io/gitpod/blob/5c84263c6f027bf3349fb476cf4c23d392db3dbc/components/dashboard/src/user-settings/Account.tsx#L163-L165) for an example) so it stays consistent with other input fields/labels.

```jsx
<InputFieldComponent label="Workspace machine type">
  <SelectWorkspaceClassComponent... />
</InputFieldComponent>
```

File: components/dashboard/src/projects/project-context.tsx
Comment (selfcontained) [inline components/dashboard/src/projects/project-context.tsx:117]: Should remove the `loading` state in this hook and just return `loading: isLoading` here (where isLoading is from the `useListProjectsQuery()`

File: components/server/src/prebuilds/bitbucket-server-app.ts
Comment (selfcontained) [inline components/server/src/prebuilds/bitbucket-server-app.ts:156]: What's the reason we need the ! here?

Comment (AlexTugarev) [reply components/server/src/prebuilds/bitbucket-server-app.ts:156]: I changed the signature of `PrebuildManager.startPrebuild` and made the `Project` arg mandatory instead of throwing an error. 
In fact `PrebuildManager.shouldPrebuild` would return false if no project is present. 

As it's raising questions here, I'll rewrite and pull out the check from `PrebuildManager.shouldPrebuild`.

Comment (AlexTugarev) [reply components/server/src/prebuilds/bitbucket-server-app.ts]: That change made sense but it was also quite a rabbit hole: https://github.com/gitpod-io/gitpod/pull/18698/commits/be8fbc60e90a0aaf5e9aae2d5820e9539a1e384e

No one wants to review this. I'll need to think how to cut this PR in pieces.

**Comment by gtsiolis** [comment]
Looking at UX changes now! 👀

File: components/server/src/prebuilds/bitbucket-app.ts
Comment (svenefftinge) [inline components/server/src/prebuilds/bitbucket-app.ts:210]: oh, how did this work before?

Comment (AlexTugarev) [reply components/server/src/prebuilds/bitbucket-app.ts:210]: Good question, @svenefftinge!

See this is the second best strategy and is only evaluated if the webhook installer isn't under the org members anymore, cf. `return { user: webhookInstaller, project }` a couple of lines above this.

But still, this is a bug. 

I wanted to extract `findProjectAndOwner` and move to `ProjectService`, but that caused so many changes, I'd rather want to make a separate PR of it.

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ProjectSettings.tsx:147]: **thought**: A checkbox might not be the best way to communicate clearly or handle properly the action here which includes also installing a webhook, but could be ok as an MVC change for now. 

I'd recommend a placeholder text as described in [EXP-589](https://linear.app/gitpod/issue/EXP-589/design-for-project-detail-page#comment-122ffba0) (internal). Then we could use a button with a proper loading state and maybe even a confirmation modal. Still fine to leave this out of the scope of this first iteration.

**Review by gtsiolis — COMMENTED** [review]
Leaving a first round of comments. 🥊 

Also scheduled a sync with @AlexTugarev taking place in the next hour to discuss more about this. 💬 

Also, thanks for the ping[[1](https://github.com/gitpod-io/gitpod/pull/18698#discussion_r1321836146)], @loujaybee.

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (AlexTugarev) [reply components/dashboard/src/projects/ProjectSettings.tsx]: Fully open to that. 👍🏻

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ProjectSettings.tsx:149]: **suggestion**: Left out the most important part, copy! 🙈 Minor rephrase to a) abstract permissions (admin, maintainer, etc) (see https://github.com/gitpod-io/gitpod/issues/16852#issuecomment-1706451086), b) mention both create+manage as you'll need permissions to be able to remove that webhook on removal, and c) make it shorter and more concise.

```suggestion
                    hint="Requires permissions to create and manage repository webhooks."
```

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (AlexTugarev) [reply components/dashboard/src/projects/ProjectSettings.tsx:147]: I wanted to start with something simple, a binary switch seems to do the job. But in fact it is a complex action, not just flipping a state. And the tasks ahead will increase the complexity behind the scene even more, which we want to be simple in front of our users. Let's see how we can get there incrementally.

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ProjectSettings.tsx:154]: **praise**: Thanks, I had a comment already in draft for this. 🌟

File: components/dashboard/src/projects/project-context.tsx
Comment (AlexTugarev) [reply components/dashboard/src/projects/project-context.tsx]: @selfcontained, I'm afraid this has some unwanted impact. When trying to load any project's page, e.g. https://at-enablep552d678423.preview.gitpod-dev.com/projects/bel-5a6d3595-5c81-4667-ab79-795b7688e817, or selecting a project from the project cards view, I'm returning back to the project cards view. I put a breakpoint here and see that `loading` from `useCurrentProject` is `false` initially. 

| | |
|--------|--------|
| Screenshot 2023-09-13 at 13 23 45 | Screenshot 2023-09-13 at 13 23 53 | 

I rely on the `isLoading` state existing in `<Projects>` to fix this, but it seems to be a the same case in other places
Screenshot 2023-09-13 at 13 31 23
Therefore, I'll go with a partial revert in this hook. Maybe we can improve this later on, WDYT?

Comment (AlexTugarev) [reply components/dashboard/src/projects/project-context.tsx]: @selfcontained, I think this might happen because we also did add `enabled: !!orgId` to `useListProjectsQuery`. The query isn't in loading state if it's not enabled, is it?

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (AlexTugarev) [reply components/dashboard/src/projects/ProjectSettings.tsx]: done

File: components/dashboard/src/data/projects/list-projects-query.ts
Comment (AlexTugarev) [reply components/dashboard/src/data/projects/list-projects-query.ts]: done

**Comment by AlexTugarev** [comment]
Quickly tested the effectiveness of the "Enable Prebuild" setting. ✔️ 
Only if enabled, push events are triggering a new prebuild. Commits are ignored after disabling prebuilds in Gitpod. 
Screenshot 2023-09-13 at 14 05 28

**Comment by svenefftinge** [comment]
reviewing this now ...

File: components/dashboard/src/projects/NewProject.tsx
Comment (svenefftinge) [inline components/dashboard/src/projects/NewProject.tsx:85]: the `!` is no longer needed as you are checking the project above. Same above.

File: components/dashboard/src/projects/ProjectListItem.tsx
Comment (svenefftinge) [inline components/dashboard/src/projects/ProjectListItem.tsx:32]: would it make sense to move this logic to a more central place, i.e. `ProjectSettings.isPrebuildsEnabled`?

Comment (AlexTugarev) [reply components/dashboard/src/projects/ProjectListItem.tsx:32]: Yes, that makes sense. I'll add this. You'll find a similar expression in prebuild-manager.ts.

File: components/gitpod-protocol/src/teams-projects-protocol.ts
Comment (svenefftinge) [inline components/gitpod-protocol/src/teams-projects-protocol.ts:18]: The backwards compatibility logic that `undefined` actually means `enabled` is very surprising. What are you concerns with usin the inverse logic internally? I.e. `prebuildsDisabled`?

**Comment by mustard-mh** [comment]
Found that `CreateProject` method on `public-api` don't have other addition settings passed [[1](https://github.com/gitpod-io/gitpod/blob/651620db9d350c23fd32c2f6f8a19bde955892d0/components/public-api-server/pkg/apiv1/project.go#L59-L64)]. Should we bring it to public-api too? Or leave it to another issue? @AlexTugarev

**Comment by AlexTugarev** [comment]
@mustard-mh, thanks for raising this!

Yes, we'd need to adjust the default value for new projects equally between the APIs! I'll take care of it. 

What's more surprising, there is no `UpdateProject` available in PAPI. 🤷🏻‍♂️ That's definitely out of scope here.

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ProjectSettings.tsx:154]: **nipick**: Minor rephrase for brevity and explicity.

```suggestion
                                Prebuilds will run for any <code>before</code> or <code>init</code> tasks.{" "}
```

examples

**Review by gtsiolis — COMMENTED** [review]
Left two more comments with copy suggestions, to match the latest designs shared in a [relevant discussion](https://gitpod.slack.com/archives/C05QPA08XAA/p1694618067747639) (internal). 🥊 

Feel free to ignore if you think they cause more confusion. 🏓

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (akosyakov) [inline components/dashboard/src/projects/ProjectSettings.tsx:86]: should here be await?

Comment (AlexTugarev) [reply components/dashboard/src/projects/ProjectSettings.tsx:86]: absolutely! thanks @akosyakov!

File: components/public-api/gitpod/experimental/v1/projects.proto
Comment (akosyakov) [inline components/public-api/gitpod/experimental/v1/projects.proto:50]: not sure it is necessary in this PR, but not so important

Comment (AlexTugarev) [reply components/public-api/gitpod/experimental/v1/projects.proto:50]: it is! 
see we're using the server API to write project settings, and PAPI to read! if this is not mapped properly, it just doesn't work with dashboard. it's out of scope for this PR to resolve this situation, but we need to act on this. not driving to the end, means we're worsening the situation, i.e. instead of an improvement this is a severe degradation, as it slows us down for each feature work in this area.

File: components/dashboard/src/projects/project-context.tsx
Comment (AlexTugarev) [reply components/dashboard/src/projects/project-context.tsx:120]: I was testing this a few times by now, and it works nicely so far, thus considering this as resolved.

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (AlexTugarev) [reply components/dashboard/src/projects/ProjectSettings.tsx:145]: We agreed to tackle this in separate task including design improvements.

Comment (gtsiolis) [reply components/dashboard/src/projects/ProjectSettings.tsx:145]: DEAL—Cross-posting [relevant discussion](https://gitpod.slack.com/archives/C05QPA08XAA/p1694680306699439?thread_ts=1694678643.951649&cid=C05QPA08XAA) (internal) and [relevant comment](https://linear.app/gitpod/issue/EXP-572/add-control-over-prebuilds-under-project-settings#comment-3430b5fd) (internal) for visibility. Cc. @loujaybee

File: components/dashboard/src/projects/NewProject.tsx
Comment (filiptronicek) [inline components/dashboard/src/projects/NewProject.tsx:84]: I would expect this button to turn on prebuilds and maybe then take me to the settings to tweak it. Otherwise, maybe replace the text with "Settings"? 

Recording 2023-09-14 at 12 39 10

**Comment by geropl** [comment]
Taking a look now :eyes:

File: components/dashboard/src/projects/NewProject.tsx
Comment (AlexTugarev) [reply components/dashboard/src/projects/NewProject.tsx:84]: Thanks @filiptronicek, you are referring to this view:
Screenshot 2023-09-13 at 15 33 27

This is change in UX we're deliberate making in a series of PRs. Decoupling webhook creation from project creation comes requires more guidance, so that one can get an idea of what to do next. @gtsiolis is going to pick this up, but in the meantime we want to ship a 🛹  so that there is at least one CTA to continue with.

File: components/dashboard/src/projects/ProjectListItem.tsx
Comment (geropl) [reply components/dashboard/src/projects/ProjectListItem.tsx:32]: (non-blocking) Following up here: Was [this method](https://github.com/gitpod-io/gitpod/blob/0492cef1498661962d160bbc17cb1da08837fc09/components/gitpod-protocol/src/teams-projects-protocol.ts#L66) meant to be used here?

**Comment by geropl** [comment]
mini-nit: Would be great if we could show the "Learn More" link also if the option is disabled, as that enables users in case they don't yet know about Prebuilds.
image

nit 2: Would be great to see some indication whether Prebuilds are enabled on this page (maybe all empty plus a central CTA "Enable Prebuilds" as we do in other places):
image

nit 3: When "Prebuilds" are disabled in the settings, I can still trigger a Prebuild manually (and I think that is good :tm: ). Still might be a question for another day/meeting/Slack thread: Have we already discussed/thought about naming it "Automatic Prebuilds"?

**Review by geropl — APPROVED** [review]
Code LGTM, tested and works :heavy_check_mark:

**Comment by AlexTugarev** [comment]
I moved the "Learn more" link, so that it's also present if prebuilds aren't enabled. 

Next: testing once more with other SCM providers before landing this.

**Comment by AlexTugarev** [comment]
Screenshot 2023-09-15 at 08 33 04

Fixed bug with BBS prebuilds via https://github.com/gitpod-io/gitpod/pull/18698/commits/418a1fa7eef7e8f1a266d0a4f4a1eda09cb27df9

**Comment by AlexTugarev** [comment]
Screenshot 2023-09-15 at 08 57 08

Fixed GitLab prebuilds in https://github.com/gitpod-io/gitpod/pull/18698/commits/453bfb3f6e251177d2738d5dd42ff9b27a28f669

**Comment by AlexTugarev** [comment]
Screenshot 2023-09-15 at 09 01 48

Checked with bitbucket.org 👍🏻

File: components/gitpod-protocol/src/teams-projects-protocol.ts
Comment (svenefftinge) [reply components/gitpod-protocol/src/teams-projects-protocol.ts:18]: If you do a data migration, then I'd love to discuss if a shape for prebuilds as suggested by @geropl would make sense. Also would be good to name the property `enabled` instead of `enable` as the later sounds like an action and not like a state.

**Pull request merged**
